File: 0701-0800/0725-Split Linked List in Parts/0725-Split Linked List in Parts.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* root, int k) {
        int len = 0;
        ListNode* head = root;
        while (head) {
            ++len;
            head = head->next;
        }
        
        int count = len / k, remain = len % k;
        vector<ListNode*> result(k);
        head = root;
        for (int i = 0; i < k && head; ++i) {
            result[i] = head;
            ListNode* prev = nullptr;
            for (int j = 0; j < count; ++j) {
                prev = head;
                head = head->next;
            }
            if (remain > 0) {
                prev = head;
                head = head->next;
                --remain;
            }
            prev->next = nullptr;
        }
        
        return result;
    }
};

File: 0701-0800/0725-Split Linked List in Parts/0725-Split Linked List in Parts.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def splitListToParts(self, root: ListNode, k: int) -> List[ListNode]:
        le = 0
        head = root
        while head:
            le += 1
            head = head.next
        count, remain = divmod(le, k)
        result = [None] * k
        head = root
        prev = None
        for i in range(k):
            if head is None:
                break
            else:
                result[i] = head
            for j in range(count):
                prev = head
                head = head.next
            if remain > 0:
                prev = head
                head = head.next
                remain -= 1
            prev.next = None
        return result
